javascript/ql/src/Security/CWE-022/ZipSlip.qhelp

```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>Extracting files from a malicious zip archive without validating that the destination file path
+is within the destination directory can cause files outside the destination directory to be
+overwritten, due to the possible presence of directory traversal elements (<code>..</code>) in
+archive paths.</p>
+
+<p>Zip archives contain archive entries representing each file in the archive. These entries
+include a file path for the entry, but these file paths are not restricted and may contain
+unexpected special elements such as the directory traversal element (<code>..</code>). If these
+file paths are used to determine an output file to write the contents of the archive item to, then
+the file may be written to an unexpected location. This can result in sensitive information being
+revealed or deleted, or an attacker being able to influence behavior by modifying unexpected
+files.</p>
+
+<p>For example, if a zip file contains a file entry <code>..\sneaky-file</code>, and the zip file
+is extracted to the directory <code>c:\output</code>, then naively combining the paths would result
+in an output file path of <code>c:\output\..\sneaky-file</code>, which would cause the file to be
+written to <code>c:\sneaky-file</code>.</p>
+
+</overview>
+<recommendation>
+
+<p>Ensure that output paths constructed from zip archive entries are validated
+to prevent writing files to unexpected locations.</p>
+
+<p>The recommended way of writing an output file from a zip archive entry is to check that
+<code>".."</code> does not occur in the path.
+</p>
+
+</recommendation>
+
+<example>
+<p>
+Here is an example of extracting an archive without validating
+filenames. If <code>archive.zip</code> contained relative paths (for
+instance, if it were created by something like <code>zip archive.zip
+../file.txt</code>) then executing this code would write to those paths.
+</p>
+
+<sample src="ZipSlipBad.js" />
+
+<p>To fix this vulnerability, we can to check that the path does not
+contain any <code>".."</code> elements in it.
+</p>
+
+<sample src="ZipSlipGood.js" />
+
+</example>
+</qhelp>
```

Comment:
For a query that has a CWE tag, you need to include at least one non-CWE reference. Then a link to the relevant CWE is automatically included in the output. It looks as if the two references used by the existing zipslip queries are probably relevant here too.

Reply:
Done.